src/cargo/util/toml/mod.rs

```diff
@@ -610,10 +645,18 @@ fn resolve_dependencies<'a>(
             }
         }
 
-        deps.insert(
-            name_in_toml.clone(),
-            manifest::InheritableDependency::Value(resolved.clone()),
-        );
+        // if the dependency is not optional, it is always used
+        // if the dependency is optional and activated, it is used
+        // if the dependency is optional and not activated, it is not used
+        let is_dep_activated =
+            !resolved.is_optional() || activated_opt_deps.contains(name_in_toml.as_str());
```

Comment:
> `name_in_toml`

Should we have a test for renamed packages to make sure the semantics of this are correct?